packages/rich-text/src/register-custom-alignment-type.js

```diff
@@ -0,0 +1,88 @@
+/**
+ * WordPress dependencies
+ */
+import { dispatch, select } from '@wordpress/data';
+
+/**
+ * Registers a new custom alignment type provided a unique name and an object defining its
+ * behavior.
+ *
+ * @param {string} name               Custom alignment name.
+ * @param {Object} settings           Custom alignment settings.
+ * @param {string} settings.blockName The block name this custom alignment will be added to the Rich Text alignment toolbar.
+ * @param {string} settings.align     The alignment setting.
+ * @param {string} settings.title     Name of the custom alignment.
+ * @param {string} settings.icon      The icon to be displayed in the alignment toolbar.
+ *
+ * @return {Object|undefined} The Custom alignment, if it has been successfully registered;
+ *                            otherwise `undefined`.
+ */
+export function registerCustomAlignmentType( name, settings ) {
+	settings = {
+		name,
+		...settings,
+	};
+
+	if ( typeof settings.name !== 'string' ) {
+		window.console.error(
+			'Custom alignment names must be strings.'
+		);
+		return;
+	}
+
+	if ( ! /^[a-z][a-z0-9-]*\/[a-z][a-z0-9-]*$/.test( settings.name ) ) {
+		window.console.error(
+			'Custom alignment names must contain a namespace prefix, include only lowercase alphanumeric characters or dashes, and start with a letter. Example: my-plugin/my-custom-alignment'
+		);
+		return;
+	}
+
+	if (
+		typeof settings.blockName !== 'string' ||
+		settings.blockName === ''
+	) {
+		window.console.error(
+			`Custom alignment block name must be a string.`
+		);
+		return;
+	}
+
+	if ( ! select( 'core/blocks' ).getBlockType( settings.blockName ) ) {
+		window.console.error(
+			'Custom alignment block "' + settings.blockName + '" must be registered.'
+		);
+		return;
+	}
+
+	if ( select( 'core/rich-text' ).getCustomAlignmentType( settings.name, settings.blockName ) ) {
+		window.console.error(
+			'Custom alignment "' + settings.name + '" is already registered.'
+		);
+		return;
+	}
+
+	if ( ! ( 'title' in settings ) || settings.title === '' ) {
+		window.console.error(
+			'The custom alignment "' + settings.name + '" must have a title.'
+		);
+		return;
+	}
+
+	if ( ! ( 'icon' in settings ) || settings.icon === '' ) {
+		window.console.error(
+			'The custom alignment "' + settings.name + '" must have an icon.'
+		);
+		return;
+	}
+
+	if ( ! ( 'align' in settings ) || settings.align === '' ) {
+		window.console.error(
+			'The custom alignment "' + settings.name + '" must have an alignment setting.'
+		);
+		return;
+	}
+
+	dispatch( 'core/rich-text' ).addCustomAlignmentTypes( settings );
+
+	return settings;
+}
```

Comment:
I'd personally love us to move away from these `register*` global functions. That said, I'd like to understand the use-case a bit better. Why do you want a new alignment option, is this meant or custom blocks essentially or for Core Blocks.

Another downside of this approach is that if you disable the custom alignment, this might create invalid blocks in some cases (depending on how the alignment is applied to the block). And I think we should discourage the extensibility APIs that lead to this.

Reply:
The use-case would be to allow custom alignment options to be added to the alignment toolbar.
The most obvious one would be `justify`, which is a standard `text-align` value, so it should be straightforward enough to add back (with my implementation).
Less obvious ones would be additional block alignments like `alignwide` and `alignfull`.

Now, this is an exploratory PR opened by someone with close-to-zero familiarity with how extending Gutenberg works, so comments like yours about the risk of breaking a block are absolutely appreciated!
This said, I didn't expect to trigger a review request to seven different people when opening the PR, or otherwise I'd have tried to describe it a bit better, and clean it up a bit more 😄

Reply:
>  didn't expect to trigger a review request to seven different people when opening the PR

I guess, it means you're lucky to have early feedback. 

About the `justify` alignment, I was thinking this is something we support by default "code-wise" but maybe just leave the UI hidden (it could be enabled using a shortcut). Is this a valid option?

Also, you're mixing both block and text alignment in the comments while it seems that at the moment, this PR is addressing text alignments only.

I wonder if this should be just another editor setting. The same way we have "colors" and "fontSizes" in the editor settings, we could include "blockAlignments" and "textAlignments". The block alignment option would be a little be redundunt with the "alignWide" setting but we can probably ensure BC. Thoughts?